### PR TITLE
Instrument eBPF for Windows platform layer

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Instrument eBPF for Windows platform layer.

Resolves: #137 

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/execution_context/ebpf_core.c [hunk 1]
```diff
@@ -182,8 +182,8 @@ _ebpf_core_protocol_load_code(_In_ const ebpf_operation_load_code_request_t* req
         if (_ebpf_core_code_integrity_state == EBPF_CODE_INTEGRITY_HYPER_VISOR_KERNEL_MODE) {
             retval = EBPF_BLOCKED_BY_POLICY;
             EBPF_LOG_MESSAGE(
-                EBPF_LEVEL_WARNING,
-                EBPF_KEYWORD_CORE,
+                EBPF_TRACLOG_LEVEL_WARNING,
```
Comment (dthaler) [inline libs/execution_context/ebpf_core.c:185]:
```suggestion
                EBPF_TRACELOG_LEVEL_WARNING,
```

File: libs/platform/ebpf_platform.h [hunk 2]
```diff
@@ -968,60 +968,60 @@ extern "C"
     void
     ebpf_trace_terminate();
 
-#define EBPF_EVENT_SUCCESS "EbpfSuccess"
-#define EBPF_EVENT_GENERIC_ERROR "EbpfGenericError"
+#define EBPF_TRACELOG_EVENT_SUCCESS "EbpfSuccess"
+#define EBPF_TRACELOG_EVENT_GENERIC_ERROR "EbpfGenericError"
 #define EBPF_EVENT_GENERIC_MESSAGE "EbpfGenericMessage"
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:973]: Should the above be EBPF_TRACELOG_EVENT_GENERIC_MESSAGE?

File: libs/platform/ebpf_platform.h [hunk 3]
```diff
@@ -968,60 +968,60 @@ extern "C"
...
+#define EBPF_TRACELOG_KEYWORD_EPOCH 0x8
+#define EBPF_TRACELOG_KEYWORD_CORE 0x10
+#define EBPF_TRACELOG_KEYWORD_LINK 0x20
+#define EBPF_TRACELOG_KEYWORD_MAP 0x40
+#define EBPF_TRACELOG_KEYWORD_PROGRAM 0x80
+
+#define EBPF_TRACELOG_LEVEL_LOG_ALWAYS WINEVENT_LEVEL_LOG_ALWAYS
+#define EBPF_TRACELOG_LEVEL_CRITICAL WINEVENT_LEVEL_CRITICAL
+#define EBPF_TRACELOG_LEVEL_ERROR WINEVENT_LEVEL_ERROR
+#define EBPF_TRACLOG_LEVEL_WARNING WINEVENT_LEVEL_WARNING
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:988]:
```suggestion
#define EBPF_TRACELOG_LEVEL_WARNING WINEVENT_LEVEL_WARNING
```

File: libs/platform/user/ebpf_handle_user.c [hunk 4]
```diff
@@ -70,7 +70,6 @@ ebpf_handle_create(ebpf_handle_t* handle, ebpf_object_t* object)
 ebpf_result_t
 ebpf_handle_close(ebpf_handle_t handle)
 {
-    EBPF_LOG_ENTRY();
```
Comment (dthaler) [inline libs/platform/user/ebpf_handle_user.c]: were these removed for perf reasons?  In other places you put a comment about to tracing due to perf.

File: libs/platform/kernel/ebpf_extension_kernel.c [hunk 5]
```diff
@@ -3,6 +3,8 @@
 
 #include "ebpf_platform.h"
 
+#pragma warning(disable : 4127) // conditional expression is constant
```
Comment (dthaler) [inline libs/platform/kernel/ebpf_extension_kernel.c:6]: Can you add a comment explaining why this line is needed?

File: libs/execution_context/ebpf_core.c [hunk 6]
```diff
@@ -174,6 +181,10 @@ _ebpf_core_protocol_load_code(_In_ const ebpf_operation_load_code_request_t* req
     if (request->code_type == EBPF_CODE_NATIVE) {
         if (_ebpf_core_code_integrity_state == EBPF_CODE_INTEGRITY_HYPER_VISOR_KERNEL_MODE) {
             retval = EBPF_BLOCKED_BY_POLICY;
+            EBPF_LOG_MESSAGE(
+                EBPF_TRACELOG_LEVEL_WARNING,
+                EBPF_TRACELOG_KEYWORD_CORE,
+                "code_type == EBPF_CODE_NATIVE blocked by EBPF_CODE_INTEGRITY_HYPER_VISOR_KERNEL_MODE");
```
Comment (dthaler) [inline libs/execution_context/ebpf_core.c:187]: independent of this PR, I think HYPER_VISOR should be changed to HYPERVISOR

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_core.c [hunk 7]
```diff
@@ -160,20 +164,27 @@ ebpf_core_terminate()
...
     ebpf_program_t* program = NULL;
     uint8_t* code = NULL;
     size_t code_length = 0;
 
     if (request->code_type == EBPF_CODE_NATIVE) {
-        if (_ebpf_core_code_integrity_state == EBPF_CODE_INTEGRITY_HYPER_VISOR_KERNEL_MODE) {
+        if (_ebpf_core_code_integrity_state == EBPF_CODE_INTEGRITY_HYPERVISOR_KERNEL_MODE) {
             retval = EBPF_BLOCKED_BY_POLICY;
+            EBPF_LOG_MESSAGE(
+                EBPF_TRACELOG_LEVEL_WARNING,
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_core.c:185]: minor: should this be EBPF_TRACELOG_LEVEL_ERROR?

File: libs/execution_context/ebpf_program.c [hunk 8]
```diff
@@ -105,27 +108,49 @@ _ebpf_program_program_info_provider_changed(
 
         ebpf_program_data_t* program_data = (ebpf_program_data_t*)provider_data->data;
         if (program_data == NULL) {
+            EBPF_LOG_MESSAGE_GUID(
+                EBPF_TRACELOG_LEVEL_ERROR,
+                EBPF_TRACELOG_KEYWORD_PROGRAM,
+                "An extension cannot have empty program_data",
+                program->parameters.program_type);
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_program.c:115]: Should we also print some information that identifies that extension? (may be provider GUID)?

File: libs/execution_context/ebpf_program.c [hunk 9]
```diff
@@ -105,27 +108,49 @@ _ebpf_program_program_info_provider_changed(
...
 
         if ((program->provider_helper_function_count > 0) &&
-            (helper_function_addresses->helper_function_count != program->provider_helper_function_count))
+            (helper_function_addresses->helper_function_count != program->provider_helper_function_count)) {
+
+            EBPF_LOG_MESSAGE_GUID(
+                EBPF_TRACELOG_LEVEL_ERROR,
+                EBPF_TRACELOG_KEYWORD_PROGRAM,
+                "A program info provider cannot modify helper function count upon reload",
+                program->parameters.program_type);
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_program.c:129]: Same as above.

File: libs/execution_context/ebpf_program.c (hunk 8 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:115]: program->parameters.program_type is the program info provider GUID.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**